File: generators/rd_waiver_generator.py
from datetime import datetime
from io import BytesIO
from PyPDF2 import PdfReader, PdfWriter
from reportlab.pdfgen import canvas
from reportlab.lib.pagesizes import letter


class RDAcceptWaiverGenerator:
    def __init__(self, template_path="templates/rd_accept_waiver.pdf"):
        self.template_path = template_path

    def generate(
        self,
        claimant: str,
        employee: str,
        case_id: str,
        rd_decision_date: str,
        current_date: str,
    ) -> tuple[str, BytesIO]:

        # filename format: RDAccept_F.Last_MM.DD.YY.pdf
        try:
            last, first = claimant.split(" ")[-1], claimant.split(" ")[0]
            short_date = datetime.now().strftime("%m.%d.%y")
            filename = f"RD_accept_waiver_{first[0]}.{last}_{short_date}.pdf"
        except Exception:
            filename = "RDAccept.pdf"

        pdf_bytes = self.draw_pdf(
            claimant, employee, case_id, rd_decision_date, current_date
        )
        return filename, pdf_bytes

    def draw_pdf(
        self, claimant, employee, case_id, rd_decision_date, current_date
    ) -> BytesIO:
        base_pdf = PdfReader(self.template_path)
        writer = PdfWriter()

        page = base_pdf.pages[0]
        overlay = self._make_overlay_page(
            claimant, employee, case_id, rd_decision_date, current_date
        )
        page.merge_page(overlay)
        writer.add_page(page)

        result = BytesIO()
        writer.write(result)
        result.seek(0)
        return result

    def _make_overlay_page(
        self, claimant, employee, case_id, rd_decision_date, current_date
    ):
        buffer = BytesIO()
        c = canvas.Canvas(buffer, pagesize=letter)
        c.setFont("Helvetica", 11)

        # Adjust these coordinates as needed
        c.drawString(410, 675, claimant)
        c.drawString(378, 662, employee)
        c.drawString(374, 647, case_id)
        c.drawString(410, 634, rd_decision_date)
        c.drawString(83, 247, claimant)
        c.drawString(315, 178, current_date)

        c.save()
        buffer.seek(0)
        return PdfReader(buffer).pages[0]

File: web/src/lib/generators/rd-waiver-generator.ts
/**
 * RD Accept Waiver Generator
 * Direct port from the deleted Python rd_waiver_generator.py to TypeScript.
 *
 * Fills templates/rd_accept_waiver.pdf ("Waiver for Recommended Decision (RD)
 * Acceptance"). Coordinates below are ported verbatim from the final Python
 * version (reportlab canvas.drawString, origin bottom-left) and should be
 * verified against a generated test PDF.
 */

import { BaseGenerator } from './base-generator';
import { ClientRecord, GeneratorResult } from './types';
import { formatDateMMDDYY, formatDateMMDDYYYY } from './utils/formatters';
import { StandardFonts } from 'pdf-lib';
import path from 'path';

export interface RDWaiverFormData {
  claimant_name: string;
  employee_name: string;
  case_id: string;
  rd_decision_date: string; // MM/DD/YYYY
  option?: '1' | '2'; // which waiver option to sign; defaults to '2'
}

export class RDWaiverGenerator extends BaseGenerator {
  constructor() {
    super('rd_accept_waiver.pdf');
  }

  async generate(
    clientRecord: ClientRecord,
    doctor: string,
    formData: RDWaiverFormData
  ): Promise<GeneratorResult> {
    const claimant = formData.claimant_name || '';
    const employee = formData.employee_name || '';
    const caseId = formData.case_id || '';
    const rdDecisionDate = formData.rd_decision_date || '';
    const currentDate = formatDateMMDDYYYY();
    const option = formData.option === '1' ? '1' : '2';

    // Each option has its own template (signature is a baked-in image that
    // must sit on the chosen option's signature line) and its own body
    // coordinates. Coordinates verified against generated test PDFs.
    if (option === '1') {
      this.templatePath = path.join(
        process.cwd(),
        'public',
        'templates',
        'rd_accept_waiver_option_1.pdf'
      );
    }
    // Option 2 keeps the constructor default (rd_accept_waiver.pdf).

    // "I, ____" name line and signature/date line differ per option because
    // the two option blocks are different heights.
    const bodyNameY = option === '1' ? 482 : 247;
    const bodyDateY = option === '1' ? 317 : 178;

    // filename format: RD_waiver_F.Last_MM.DD.YY.pdf
    const shortDate = formatDateMMDDYY();
    let filename = 'RDWaiver.pdf';
    const parts = claimant.trim().split(/\s+/);
    if (parts.length >= 2) {
      const first = parts[0];
      const last = parts[parts.length - 1];
      filename = `RD_waiver_${first[0]}.${last}_${shortDate}.pdf`;
    }

    // Load template and generate PDF
    const pdfDoc = await this.loadTemplate();
    const font = await pdfDoc.embedFont(StandardFonts.Helvetica);

    const page = pdfDoc.getPages()[0];
    page.setFont(font);
    page.setFontSize(11);

    // Header block coordinates are identical for both options.
    this.drawText(page, caseId, { x: 410, y: 675, size: 11 });
    this.drawText(page, employee, { x: 378, y: 662, size: 11 });
    this.drawText(page, claimant, { x: 374, y: 647, size: 11 });
    this.drawText(page, rdDecisionDate, { x: 410, y: 634, size: 11 });
    // Body name + date land on the selected option's block.
    this.drawText(page, claimant, { x: 83, y: bodyNameY, size: 11 });
    this.drawText(page, currentDate, { x: 315, y: bodyDateY, size: 11 });

    const pdfBytes = await pdfDoc.save();

    return {
      filename,
      pdfBytes: Buffer.from(pdfBytes),
    };
  }
}
